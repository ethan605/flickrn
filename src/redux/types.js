export const PREFIXES = {
  FLICKR: 'flickr/',
  SHARED: 'shared/',
};

export const FLICKR = {
  CLEAR_PHOTO_INFO: 'flickr/CLEAR_PHOTO_INFO',
  GET_PHOTO_INFO: 'flickr/GET_PHOTO_INFO',
  SEARCH_PHOTOS: 'flickr/SEARCH_PHOTOS',
};

export const SHARED = {
};
