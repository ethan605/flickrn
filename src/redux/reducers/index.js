import photoInfo from './photoInfo';
import photos from './photos';

export { photoInfo, photos };
