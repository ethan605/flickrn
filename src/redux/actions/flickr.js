import _ from 'lodash';
import { createAction } from 'redux-actions';
import { FLICKR } from 'src/redux/types';

// Constants
import { API_CONFIGS } from 'src/constants';

const { flickr: { apiKey } } = require('src/data/configs.json');
const { PAGINATION } = API_CONFIGS;

const DEFAULT_PARAMS = {
  api_key: apiKey,
  format: 'json',
  nojsoncallback: 1,
  per_page: PAGINATION,
};

export const clearPhotoInfo = createAction(FLICKR.CLEAR_PHOTO_INFO);

export function getPhotoInfo({ id, secret }) {
  const action = createAction(FLICKR.GET_PHOTO_INFO);

  return dispatch => {
    const request = {
      params: {
        ...DEFAULT_PARAMS,
        secret,
        method: 'flickr.photos.getInfo',
        photo_id: id,
      },
    };

    return dispatch(action({ request }));
  };
}

export function searchPhotos({ keyword, reloading }) {
  const action = createAction(FLICKR.SEARCH_PHOTOS);

  return (dispatch, getState) => {
    const { photos: { currentPage } } = getState();

    const method = _.isEmpty(keyword) ? 'getRecent' : 'search';
    
    const request = {
      params: {
        ...DEFAULT_PARAMS,
        method: `flickr.photos.${method}`,
        page: reloading ? 1 : currentPage + 1,
        text: keyword,
      },
    };

    return dispatch(action({ request }));
  };
}
