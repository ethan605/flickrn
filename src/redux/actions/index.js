import * as flickr from './flickr';

/* eslint-disable import/prefer-default-export */

export { flickr };
