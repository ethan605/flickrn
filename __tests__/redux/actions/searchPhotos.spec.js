// Helpers
import { DEFAULT_PARAMS, buildAxiosActions, mockAxios } from '__tests__/helpers/mockApiRequests';
import buildMockStore from '__tests__/helpers/buildMockStore';

// Redux
import { flickr } from 'src/redux/actions';

describe('redux/actions/searchPhotos', () => {
  const store = buildMockStore({ authorized: true });

  const keyword = 'random keyword';
  const mockSuccess = require('__tests__/fixtures/searchPhotos/success.json');
  const mockError = require('__tests__/fixtures/searchPhotos/error.json');

  const requestAction = {
    type: 'flickr/SEARCH_PHOTOS',
    payload: {
      request: {
        params: {
          ...DEFAULT_PARAMS,
          method: 'flickr.photos.search',
          page: 1,
          text: keyword,
        },
      },
    },
  };

  const { photos } = mockSuccess;
  const { errorAction, successAction } = buildAxiosActions(requestAction, { photos }, mockError);

  /* eslint-disable newline-per-chained-call */
  mockAxios
    .onGet('').replyOnce(200, mockSuccess)
    .onGet('').replyOnce(200, mockError);
  /* eslint-enable newline-per-chained-call */

  beforeEach(() => store.clearActions());

  test('success', async () => {
    await store.dispatch(flickr.searchPhotos({ keyword, reloading: true }));
    const actions = store.getActions();
    expect(actions.length).toEqual(2);
    expect(actions).toEqual([requestAction, successAction]);
    expect(actions).toMatchSnapshot();
  });
  
  test('error', async () => {
    await store.dispatch(flickr.searchPhotos({ keyword, reloading: true }));
    const actions = store.getActions();
    expect(actions.length).toEqual(2);
    expect(actions).toEqual([requestAction, errorAction]);
    expect(actions).toMatchSnapshot();
  });
});
