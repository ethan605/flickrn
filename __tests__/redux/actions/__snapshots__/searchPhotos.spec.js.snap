// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/actions/searchPhotos error 1`] = `
Array [
  Object {
    "payload": Object {
      "request": Object {
        "params": Object {
          "api_key": "92ab2a894f0effc28f920362ff52fc2d",
          "format": "json",
          "method": "flickr.photos.search",
          "nojsoncallback": 1,
          "page": 1,
          "per_page": 20,
          "text": "random keyword",
        },
      },
    },
    "type": "flickr/SEARCH_PHOTOS",
  },
  Object {
    "error": Object {
      "message": "Random flickr.photos.search error",
      "response": Object {
        "data": Object {
          "code": 123,
          "message": "Random flickr.photos.search error",
          "stat": "fail",
        },
      },
    },
    "meta": Object {
      "previousAction": Object {
        "payload": Object {
          "request": Object {
            "params": Object {
              "api_key": "92ab2a894f0effc28f920362ff52fc2d",
              "format": "json",
              "method": "flickr.photos.search",
              "nojsoncallback": 1,
              "page": 1,
              "per_page": 20,
              "text": "random keyword",
            },
          },
        },
        "type": "flickr/SEARCH_PHOTOS",
      },
    },
    "type": "flickr/SEARCH_PHOTOS:ERROR",
  },
]
`;

exports[`redux/actions/searchPhotos success 1`] = `
Array [
  Object {
    "payload": Object {
      "request": Object {
        "params": Object {
          "api_key": "92ab2a894f0effc28f920362ff52fc2d",
          "format": "json",
          "method": "flickr.photos.search",
          "nojsoncallback": 1,
          "page": 1,
          "per_page": 20,
          "text": "random keyword",
        },
      },
    },
    "type": "flickr/SEARCH_PHOTOS",
  },
  Object {
    "meta": Object {
      "previousAction": Object {
        "payload": Object {
          "request": Object {
            "params": Object {
              "api_key": "92ab2a894f0effc28f920362ff52fc2d",
              "format": "json",
              "method": "flickr.photos.search",
              "nojsoncallback": 1,
              "page": 1,
              "per_page": 20,
              "text": "random keyword",
            },
          },
        },
        "type": "flickr/SEARCH_PHOTOS",
      },
    },
    "payload": Object {
      "photos": Object {
        "page": 1,
        "pages": 39955,
        "perpage": 5,
        "photo": Array [
          Object {
            "farm": 5,
            "id": "40107995774",
            "isfamily": 0,
            "isfriend": 0,
            "ispublic": 1,
            "owner": "152508136@N06",
            "secret": "7ed5905c16",
            "server": "4778",
            "title": "üê±üê∂üê∂ üíñ",
          },
          Object {
            "farm": 5,
            "id": "26945278248",
            "isfamily": 0,
            "isfriend": 0,
            "ispublic": 1,
            "owner": "128512949@N07",
            "secret": "9d9fc82c4c",
            "server": "4777",
            "title": "@Cgshapes , @Ella , @Fashionnatic , @Serendipity , @Armonia",
          },
          Object {
            "farm": 5,
            "id": "40772774262",
            "isfamily": 0,
            "isfriend": 0,
            "ispublic": 1,
            "owner": "52361229@N06",
            "secret": "24942a6a29",
            "server": "4771",
            "title": "These Guys starting 3 days early.....disclaimer: no pets or dogs were hurt or got intoxicated during the making of this image.",
          },
          Object {
            "farm": 5,
            "id": "38999884280",
            "isfamily": 0,
            "isfriend": 0,
            "ispublic": 1,
            "owner": "162254693@N02",
            "secret": "0e07ee8b6e",
            "server": "4784",
            "title": "Corgi king",
          },
          Object {
            "farm": 5,
            "id": "39914583135",
            "isfamily": 0,
            "isfriend": 0,
            "ispublic": 1,
            "owner": "113454068@N06",
            "secret": "25e007537b",
            "server": "4783",
            "title": "Sleeping Chorgi",
          },
        ],
        "total": "199773",
      },
    },
    "type": "flickr/SEARCH_PHOTOS:SUCCESS",
  },
]
`;
